MSOutput nonRelVal - two run modes for MSOutput thread

todor-ivanov opened the pull request:
Fixes #9609 

#### Status
Ready

#### Description
Changes for the `MSOutput` Module needed for `nonRelVal` Workflows. For more details about the requirements to the module see the Description in the Issue. For the Current PR  read bellow.

This PR should provide the following functionality:
   - Fetch all Request Records in status `completed`, `announce`, `closed-out` from ReqMgr 
   - Create Data transfer subscriptions to the relevant Data Management System (DDM, Phedex, Rucio)
   - Implement local book keeping for transfers subscribed

It should work in two modes Producer an Consumer:
   - The Producer is a single thread fetching data from Reqmgr API and uploading ti to MongoDB
   - The Consumers are N threads to fetch records from MongoDB and take the relevant actions about making the the needed output data placement transfers, taking into account  the workflow type (`RelVal` | `NonRelval`) and eventual campaign/workflow  output data mapping.   

In order to reduce Memory footprint and increase scalability, the `Producer` is split in two basic components:
   - A Functional Pipeline to apply all the changes and transformations to the documents one step at a time and upload them to MongoDB.
   - A Document Streamer (not yet fully implemented, may come with another PR, though) to feed the pipeline either a single document at a time or in strides of N documents at a time (in order to reduce communication with database in the future).

*NOTE:* A similar approach is intended for the `Consumer` too.

*NOTE*: The current PR is related to `NonRelVal` workflows only,  as well as to all the auxiliary code changes and modules needed to implement the service architecture explained above.

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs

#### External dependencies / deployment changes
NO

todor-ivanov committed:
MSOutput nonRelVal - two run modes for MSOutput thread

todor-ivanov committed:
MongoDBUploader function and thread

Adding logger && self references

Move Pipeline under Utils && fix broken imports

Changing hardcoded MongoDb configuration (temporary - for testing purposes).

Fix bad arguments reference

Fix empty results from getRequestRecords && missing list_database_names() for older pymongo version

Auxiliary Functor Class for the Pipeline

Move db connection to __init__ && Read Server and Port from msConfig.

Changing thread names to MSOutput{Producer,Consumer}

Enforce self consistency check && Add checks for mandatory fields && Add setRelVal method.

Set docCleaner function at the end of the pipes

Renaming mongoDBUploader to msOutputProducer && adding a new step in the pipeline for updating document information from campaign info

todor-ivanov commented:
While working with this PR I explored some kind of a different approach. In order to improve the scalability and to avoid huge memory footprint, I decided to implement the service in a series of function calls that  are happening through a functional pipeline. The pipeline is feed from a document streamer (generator) instead of a huge in memory encoded Json file. At the end of the pipeline  the objects are either destroyed or just left to be deleted by the Garbage Collector because they go out of scope of the functions from the pipeline. 

Here [1] is the final benchmark for the MongoDBUploader working on my laptop and reading from a 69MB json file with requests stored on my hard disk and a remote MongoDB Server (running on a VM in my laptop again). The indexes in the collections are created on the "RequestName" key. And all the documents are unique on this key. 

[1]

```
In [276]: %time %mprun -f mongoDBUploader mongoDBUploader()
len result: 7361
counter: 7361
Filename: /home/tivanov/Projects/WMCoreDev.d/test/ms-output-test/ms_output_test.py
Line #    Mem usage    Increment   Line Contents
================================================
    87     97.9 MiB     97.9 MiB   def mongoDBUploader():
    88                                 """
    89                                 A top level function to drive the upload of all the documents to MongoDB
    90                                 """
    91                                 # TODO:
    92                                 #    to set a destructive function at the end of the pipeline
    93                                 # NOTE:
    94                                 #    To discuss the collection names
    95     97.9 MiB      0.0 MiB       counter = 0
    96     97.9 MiB      0.0 MiB       msOutIndex = IndexModel('RequestName', unique=True)
    97     97.9 MiB      0.0 MiB       msOutputDB = MongoDB('msOutputDB',
    98     97.9 MiB      0.0 MiB                            create=True,
    99                                                      collections = [
   100     97.9 MiB      0.0 MiB                                ('msOutRelValColl', msOutIndex),
   101     97.9 MiB      0.0 MiB                                ('msOutNonRelValColl', msOutIndex)]).msOutputDB
   102                             
   103     97.9 MiB      0.0 MiB       msOutRelValColl = msOutputDB['msOutRelValColl']
   104     97.9 MiB      0.0 MiB       msOutNonRelValColl = msOutputDB['msOutNonRelValColl']
   105     97.9 MiB      0.0 MiB       msPipelineRelVal = Pipeline([docTransformer,
   106     97.9 MiB      0.0 MiB                                    (docUploader, msOutRelValColl)])
   107     97.9 MiB      0.0 MiB       msPipelineNonRelVal = Pipeline([docTransformer,
   108     97.9 MiB      0.0 MiB                                       (docUploader, msOutNonRelValColl)])
   109    210.9 MiB    113.0 MiB       for _, request in docStreamer():
   110    210.9 MiB      0.0 MiB           counter += 1
   111    210.9 MiB      0.0 MiB           if 'SubRequestType' in request.keys() and 'RelVal' in request['SubRequestType']:
   112    210.9 MiB      0.0 MiB               msPipelineRelVal.run(request)
   113                                     else:
   114    210.9 MiB      0.0 MiB               msPipelineNonRelVal.run(request)
   115                                     # if counter == stride:
   116                                     #     break
   117    103.6 MiB      0.0 MiB       print('counter: %s' % counter)
CPU times: user 11.7 s, sys: 299 ms, total: 12 s
Wall time: 14.3 s
In [277]: 
In [277]: 
In [277]: msOutputDB['msOutRelValColl'].count_documents({})
Out[277]: 124
In [278]: msOutputDB['msOutNonRelValColl'].count_documents({})
Out[278]: 7237
```

vkuznet commented:
Todor, does it means that the memory footprint comes from `docStreamer()`? And, can you explain why it takes 113MiB? Is it because of it wraps the data or it is size of the data? If `docStreamer()` is a generator it should consume only memory needed for a single document, otherwise, it may have wrong logic to handle the data.

todor-ivanov commented:
@vkuznet - Yes I can explain it. It is because I am faking the docStreamer, since we do not have a real one yet. I am actually loading the whole data in memory from a file at once, just before I call the streamer, and I then stream (destructively pop)  a document at a time from there. On the other side I was expecting to see was a decrease of memory in that line on every second cycle.... but unfortunately the report from the 'meomory_profiler' module reports values which are somehow  accumulated thorough out of the whole run, but does not break them per cycle.

amaltaro commented:
Todor, don't take it as a review because I simply "glanced" at this PR. I have a few comments though:
1. please elaborate on the initial description. This provides a reasonable amount of changes, and simply pointing to the GH issue is not efficient and clear
2. how about we rename the MSOutput threads to: MSOutputProducer, MSOutputConsumer?
3. I like very much the MongoDB module created under WMCore/Database (I haven't checked any implementation details though)
4. can you please also clarify what are the benefits of the `Pipeline` module? Given that it does not depend on any WMCore modules, it might make sense to move that under `Utils` instead. You can expand a bit on it in the initial description as well

todor-ivanov commented:
Hi @amaltaro Thanks  for the comments:

> 1. please elaborate on the initial description.

Done

> 2. how about we rename the MSOutput threads to: MSOutputProducer, MSOutputConsumer?

A deal We got!

> 3. MongoDB Module

MongoDB has the so called Lazy database and collection creation - which happens late on the first document uploading, while in the meantime all the objects and sessions are created and inactive. So ti basically hides from you almost any error that my happen in the background until you actually hit it.This module is supposed to check everything in advance and also to create the needed indexes mostly because if we start with collections created by default we will not have the document index and the `unique` property for the index - which assures us we do not grow the database infinitely and we keep a closed set of documents in the system (regardless of the changes of the status of the workflow that this documents describe)

> 4. benefits of the Pipeline module?

So basically avoids objects reassignment of type `x = x+2` which happens automatically internally in the pipeline. It also (supposedly - if we implement it correctly) should hold the objects in the memory only until they are in the pipeline. Once they are out we can either explicitly delete them at the end of the pipeline (I tried putting `del` there as a last call in the pipeline but did not work though :) )or just let them get out of scope (if  we keep everything enclosed only in the calling function scope), and then let the Garbage collector do its job. It also makes the code quite readable to me - you just create a set of pipelines with the proper actions you want to happen with the object, and on top of it you just create a simple dispatcher (in my case even a simple `if` fork works ) that throws the object in the relevant pipe. and you let the rest of the machinery do it's job ;)

Commit:
Introducing a simple streamer.

todor-ivanov committed:
Fix broken chain in the pipeline - No document returned from docUpdater

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+    database) documents. And it must have all the methods needed to set or update
+    various fields in the object. So that we assure the uniformity of all documents.
+
+    Template format:
```
This template is not compliant with what has been discussed in the issue #9610, such as:
* "WorkflowName" reads "RequestName" over here. It's fine IMO, but this things need to be fed back to the issue where the requirements are laid out (update the issue if you keep it as is) 
* "RequestStatus" needs to go away
* "transferStatus" reads "isTaken" over here. Please update the GH issue with this parameter.
* "outputDatasets" reads "OutputDatasets" over here (capital). Please update the code here.

I see a clash between "isTaken" and "transferStatus", we don't need both arguments. So we need to pick one and use it for the consumer lookup. I don't have any preference, but the GH issue needs to reflect these changes.

todor-ivanov replied:
I will update the GH  issue. I basically changed the names so we can have as much overlap with Reqmgr as possible and also - I am actually automatically copying them from the initial document to the one which is about to be uploaded in MongoDB. look at this [1]
   * `RequestName` is an example of this exact thing
   * `RequestStatus` will go away maybe next month - this is put there temporarily wont be used for any login regarding the subscritopns
   * `OutputDatasets`   Same as `RequestName`
   * `transferStatus` does not overlap `isTaken` - one is the status of the Transfer at the Data Management system, the other is the flag if the document is taken from on of the threads of the MSOutput Module and is under evaluation - it is put there in order to avoid race conditions between the threads.
 
And overall about the initial capital letters - l would suggest that similar to Reqmgr we  capitalize them all here too.   

[1]
https://github.com/dmwm/WMCore/blob/eb2edafab1639a248d1771a1a344e0317710355e/src/python/WMCore/MicroService/Unified/MSOutputTemplate.py#L94-L101

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+    "numberOfCopies": integer
+    }
+    """
+    # TODO:
```
I think we can default the id to request['RequestName'], fail if such key does not exist.

todor-ivanov replied:
It happens automatically because of the piece of code I just pasted in the previous comment

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+        super(MSOutputTemplate, self).__init__(**kwargs)
+        # super(MSOutputTemplate, self).__init__()
+
+        docTemplate = [
```
See my comments from the docstring above. These need to be updated.
BTW, can we change the data type from `unicode` to `str`? Unicode does not exist in the py3 world.

todor-ivanov replied:
Actually, No from Mongo they come as `unicode` - it was before a `str` here, but it was breaking badly because of that so I had to change it to `unicode`. I made a comment in the code regarding this exactly [1]

[1]
https://github.com/dmwm/WMCore/blob/eb2edafab1639a248d1771a1a344e0317710355e/src/python/WMCore/MicroService/Unified/MSOutputTemplate.py#L143-L144

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+            ('_id', None, unicode),
+            ('RequestName', None, unicode),
+            ('RequestStatus', None, unicode),
+            ('creationTime', None, Timestamp),
```
I think you already explained it to me. Can you please remind me why these timestamps are in different format? I'd make them consistent.

todor-ivanov replied:
I was actually explaining why I like the `Timestamp` from Bson more than the standard time format which is basically `float`. But mostly - because this is a format internally used by Mongo to represent timestamps so I would rather stick to it from the very beginning of the document creation. 
Aaand... about why the format of `lastupdated` is still a float the answer quite trivial - I forgot to change it ;)
Fixing it now

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+            # the document passed and fill them into internal document so that they
+            # can pass the needed checks later, throw the unneeded/unrecognised
+            # key/values from the passed document
+            for key in map(lambda tup: tup[0], docTemplate):
```
Please check the pylint report on the map function

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+            # key/values from the passed document
+            for key in map(lambda tup: tup[0], docTemplate):
+                if key in doc.keys():
+                    myDoc[key] = deepcopy(doc[key])
```
If we will operate on the mongo doc, any reason not to pass the value by reference?

todor-ivanov replied:
Yes, one  reason and it is quite solid - because of that line here [1]. I am eager to drop everything that I do not need from that huge document that comes  from Reqmgr. And with that line there things will vanish from the other document too ,which at that moment is not uploaded to MongoDB yet.

[1]
https://github.com/dmwm/WMCore/blob/eb2edafab1639a248d1771a1a344e0317710355e/src/python/WMCore/MicroService/Unified/MSOutput.py#L346

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+        if self._checkValid(myDoc, throw=True):
+            self.update(myDoc)
+
+    def checkAttr(self):
```
is this method ever called?

todor-ivanov replied:
It is not called anywhere in the code, but I put it here because while developing and testing it turned out to be quite useful. Especially when I mess up with the documents directly in iPython and try to break stuff intentionally, at some point I need to be able to see what I am doing so I just call this one to provoke a self consistency check of the object. This works quite well.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+
+    def setRelVal(self, isRelVal):
+        myDoc = {'isRelVal': isRelVal}
+        if self._checkAttr(self.docTemplate, myDoc, throw=False, update=False, **myDoc):
```
Do I read it right that it validates the whole document before it updates the `isRelVal` attribute?

todor-ivanov replied:
No, you are wrong. (Sorry to say that :) ) It is actually checking only the portion that  is about to be updated. And in this case this is: `myDoc = {'isRelVal': isRelVal}` . Notice in the call:
```
if self._checkAttr(self.docTemplate, myDoc, throw=False, update=False, **myDoc):
```
It checks myDoc  against itself without updating values and without throwing an exception - just getting the `bool` value from passing everything though the `self.docTemplate`.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+            return True
+        return False
+
+    def setWflowType(self):
```
Unless you have a use case to use it in the next weeks, please remove. I see no use for it.

todor-ivanov replied:
I am about to clean those in one of the next commits Once I am done with the `Consumer` part.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+    def updateStatus(self):
+        pass
+
+    def strip(self):
```
Unless you have a use case to use it in the next weeks, please remove. I see no use for it.

todor-ivanov replied:
Same as  above.

todor-ivanov replied:
Removed

amaltaro reviewed (changes requested):
Todor, I made a very long review (so long that the MSOutputTemplate wasn't reviewed very carefully).
I placed a bunch of comments in the code. If some of them are important but won't be worked in this PR, please leave them as unresolved such that we can get back to this PR.

todor-ivanov commented on src/python/Utils/Pipeline.py:
```diff
@@ -0,0 +1,113 @@
+from functools import reduce
+
+
+class Functor:
```
Done!

todor-ivanov commented on src/python/Utils/Pipeline.py:
```diff
@@ -0,0 +1,113 @@
+        return self.func(obj, *self.args, **self.kwargs)
+
+
+class Pipeline:
```
Done

todor-ivanov commented on src/python/Utils/Pipeline.py:
```diff
@@ -0,0 +1,113 @@
+    A simple Functional Pipeline Class: applies a set of functions to an object,
+    where the output of every previous function is an input to the next one
+    """
+    # DONE:
```
Done

todor-ivanov commented on src/python/Utils/Pipeline.py:
```diff
@@ -0,0 +1,113 @@
+    # DONE:
+    #   To implement passing arguments to the functions - it will allow us to
+    #   keep the object creation outside of the pipeline
+    def __init__(self, funcLine=[]):
```
Done

todor-ivanov commented on src/python/Utils/Pipeline.py:
```diff
@@ -0,0 +1,113 @@
+    #   keep the object creation outside of the pipeline
+    def __init__(self, funcLine=[]):
+        """
+        :obj: The object to be passed through the pipeline
```
Yes indeed. Fixed!

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+                 server=None,
+                 port=None,
+                 create=False,
+                 collections=[],
```
Fixed

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+                 testIndexes=False,
+                 logger=None):
+        """
+        :databases:   A list of databases to connect to
```
No, you did not. In the previous implementation it was a list indeed.

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+                 collections=[],
+                 testIndexes=False,
+                 logger=None):
+        """
```
besides  the 'list' to 'string' mismatch you mention bellow (now fixed, the rest is still valid.

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+        :databases:   A list of databases to connect to
+        :create:      A flag to trigger a database creation (if missing) during
+                      object construction, together with collections if present.
+        :collections: A list of tuples describing collections with indexes -
```
having them defined as a tuple will allow us to make the following call to the future `self._inedxTest` function :
```
self._indexTest(*collection)
```

Whose signature BTW in the current version is wrong it should be:
```
def _indexTest(self, collection, index)
```

To me the tuple looks much easier and flexible to use. But if you insist I can change to dictionary though.
I am leaving this unresolved for the moment, would like to hear back from you.

todor-ivanov replied:
I actually had to fix it last night because pylint was complaining and I did it wrong. Now it is done properly. Thanks for bringing it up again.

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+        :testIndexes: A flag to trigger index test and eventually to create them
+                      if missing (TODO)
+        """
+        # DONE:
```
Removed

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+            self.client = MongoClient(self.server, self.port)
+            self.client.server_info()
+        except Exception as ex:
+            msg = "ERR: could not connect to MongoDB server: %s\n%s" % (self.server, str(ex))
```
Fixed. I am leaving only one explicit `ERR:` prefix on line: [1]. Because there  we are not logging but just throwing an exception. So from this place here there is no way to know which or what would be the top level  object to catch that and how would it treat it (it eventually may or may not use logger.erroro() to log it etc.). 


[1]
https://github.com/dmwm/WMCore/pull/9691/commits/eb2edafab1639a248d1771a1a344e0317710355e#diff-4ec8a9000f12097e2fd4dedc005dc385R89

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+
+    def _collCreate(self, coll, db):
+        """
+        :coll: A tuple describing one collection with indexes -
```
Done

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+
+               https://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.create_index
+        """
+        # DONE:
```
Removed

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+        collIndexes = list(coll[1:])
+        try:
+            self.client[db].create_collection(collName)
+        except errors.CollectionInvalid:
```
I was just trying to avoid having this message at the beginning of every polling cycle of the service. But Ok, lets log it.  
Done now

todor-ivanov replied:
Done: See https://github.com/dmwm/WMCore/issues/9711

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+            # this error is thrown in case of an already existing collection
+            pass
+        if collIndexes:
+            # DONE:
```
Removed

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+                raise ex
+
+    def _dbTest(self, db):
+        # Test connection (from mongoDB documentation):
```
Done

todor-ivanov commented on src/python/WMCore/Database/MongoDB.py:
```diff
@@ -0,0 +1,156 @@
+        except errors.InvalidName as ex:
+            msg = "ERR: Could not connect to a missing MongoDB databases: %s\n%s" % (db, str(ex))
+            self.logger.error(msg)
+            if self.create:
```
If you are talking about having the database object creation fully outside the MSOutput module and put it somewhere on the top during the thread creation.... I can see the benefit from that... I agree. And even I already made some effort to move it as much as possible to the top of the initialization sequence (at the beginning it was inside the Producer function call), but I simply did not want to go out of the MSOutput.py boundaries for the time being. I would say, lets go with that as it is right now (we can berry a repetition of object recreation once per pooling cycle) and we create a separate issue and pull this piece out of the module next month, because it will need additional testing.  

In case I misunderstood you please cerrect  me :)

todor-ivanov replied:
Done: See https://github.com/dmwm/WMCore/issues/9711

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSManager.py:
```diff
@@ -94,17 +94,28 @@ def __init__(self, config=None, logger=None):
+                                                          reqStatus,
+                                                          self.msConfig['interval'],
+                                                          self.logger))
+            self.logger.debug(
```
Alan, Those are repetitions from lines: [1] && [2]. I was following the same stile and logging level as for them. If you think we should chage I would rather change it for all of them.  I am leaving this one unresolved too - would like to hear from you too.  

[1]
https://github.com/dmwm/WMCore/blob/eb2edafab1639a248d1771a1a344e0317710355e/src/python/WMCore/MicroService/Unified/MSManager.py#L92

[2]
https://github.com/dmwm/WMCore/blob/eb2edafab1639a248d1771a1a344e0317710355e/src/python/WMCore/MicroService/Unified/MSManager.py#L80

todor-ivanov replied:
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSManager.py:
```diff
@@ -94,17 +94,28 @@ def __init__(self, config=None, logger=None):
+
+            self.msOutputProducer = MSOutput(self.msConfig, logger=self.logger,
+                                             mode='MSOutputProducer')
+            thname = 'MSOutputProducer'
```
And this is an idea I like very much :). Thanks
Done now

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSManager.py:
```diff
@@ -94,17 +94,28 @@ def __init__(self, config=None, logger=None):
+            thname = 'MSOutputProducer'
+            self.outputProducerThread = start_new_thread(thname, daemon,
+                                                         (self.outputProducer,
+                                                          reqStatus,
```
No it is not! It id not used anywhere. I just fetched it only because I wanted to experiment with updates and synchronization of some dynamic values between Reqmgr and MongoDB (some that are  supposed to change in time). Because all the rest of the values that we are collection right now are static throughout the whole workflow life time. 
We can remove it safely in the future, but for the moment I would like to keep it only to see how it goes and how should we deal with such a synchronization in the future.

todor-ivanov replied:
Almost... :) I am pretty sure they will get aligned at some point for sure. It simply would have taken some time for them to get in sync. And that's what I wanted to test, but anyway, I will leave it for some other time. Removing it from the document for now.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -44,24 +59,70 @@ def __init__(self, msConfig, logger=None):
+                ('msOutRelValColl', msOutIndex),
+                ('msOutNonRelValColl', msOutIndex)]}
+
+        self.msOutDB = MongoDB(**msOutDBConfig).msOutDB
```
Oook, this one is tricky. And I agree it is not perfect, I do not like it either. So things start from how Mongo is creating a database session: [1]. Nothing wrong with that of course, but as you may have already noticed, the `MongoDB` class does not inherit directly from `pymongo.MongoClient` (which I should have done from the very beginning ... but any way). So now the actual database object is attached to the `MongoDB` instance as an attribute here [2]. (and because of the aforementioned lack of inheritance, even the client call is not just `self['SomedbName']` as one should expect, but it is `self.cleint['SomedbName']` instead - because the client is also attached to the object as an attribute). So this is how I ended up with that call:
```
dbObject = MongoDB(database='dbObject').dbObject
```
*NOTE:* Please notice - it will break if the database string in this call is not the same as the variable name - I know this is horrible.

I could have updated the object's __dict__ also so we can have calls similar to the the one that are implemented in the MongoClient. But it would just have lead to repetitions in the call. It would look something like:

```
 dbObject = MongoDB(database='dbObject')['dbObject']
```
As you see - not much better. And internaly in the class the thing would look even worse. 

My Opinion is that this 'MongoDB' module, even as good as it is right now, will still need a refactoring in the next cycle. So an issue combining this one here and the a fore mentioned 'moving the Database instance on top of the thread' is a must for the next month I would say.

[1]
https://api.mongodb.com/python/3.4.0/tutorial.html#getting-a-database 

[2]
https://github.com/dmwm/WMCore/blob/eb2edafab1639a248d1771a1a344e0317710355e/src/python/WMCore/Database/MongoDB.py#L123

todor-ivanov replied:
I would rather go for making our Class inheriting from Mongo Client and test if everything works properly which would make us quite aligned with the behavior of the upstream provided tools and just enriching the object with the methods we need for our specific task.
And yes, we'd better follow this on a separate issue - I am refering it here again: https://github.com/dmwm/WMCore/issues/9711

vkuznet replied:
Todor, I'm honestly lost in all these changes, but please do not make WMCore classes inherit from 3rd party APIs as you have no control over 3rd party libraries/APIs, and if they change their class behavior your class will misbehave as well without any notice.

I also don't see any point wrapping out Mongo APIs with WMCore APIs for the same reason. And, I rather will use 3rd party APIs directly which can be easy to fix in a long run.

amaltaro replied:
Valentin has a good point about inheriting from a 3rd party library.
About the wrapper, I'm all in favor of it because then we can put all the extra necessary logic/exception handling/etc around the client APIs and keep a consistent usage of them.

todor-ivanov replied:
Thanks @vkuznet and @amaltaro  I can see your point and Agree with that. I wont make it inherit from the Mongo Client, but will try few other things in order to the object construction.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -44,24 +59,70 @@ def __init__(self, msConfig, logger=None):
+                ('msOutNonRelValColl', msOutIndex)]}
+
+        self.msOutDB = MongoDB(**msOutDBConfig).msOutDB
+        self.msOutRelValColl = self.msOutDB['msOutRelValColl']
```
I am not sure I fully understand your comment here. Did you mean something like that: 
```
self.msOutRelvalCollName = 'msOutRelValColl'
self.msOutRelValColl = self.msOutDB[self.msOutRelvalCollName]

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -44,24 +59,70 @@ def __init__(self, msConfig, logger=None):
         :return: summary
         """
+
+        # DONE:
```
Removed

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -75,35 +136,60 @@ def execute(self, reqStatus):
-            msg = "Unknown exception while fetching requests from ReqMgr2. Error: %s", str(err)
-            self.logger.exception(msg)
-            self.updateReportDict(summary, "error", msg)
+        if self.mode == 'MSOutputProducer':
```
I agree - it will look much nicer. I will make a `TODO:` for the moment :)

todor-ivanov replied:
Actually I just split them and it is done now.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutputTemplate.py:
```diff
@@ -0,0 +1,228 @@
+import time
```
Done

todor-ivanov commented:
Hi Alan @amaltaro, Thank you very much for the good review. I did leave some of the conversations unresolved Please take a look and comment if you think any of those need to be fixed before we merge this PR. And actually for some of them I do need your feedback/opinion.I am now pushing one commit with few of the changes requested in your review.  And few more commits to come later tonight. 
Thanks again.

todor-ivanov committed:
First review iteration.

amaltaro commented:
Thanks, Todor. I have unresolved conversations that need any sort of follow up. I might only get to those tomorrow though. Let's try to get this PR merged by my Tuesday morning.

todor-ivanov committed:
MSOutputConsumer mode - main functionality.

Handling exceptions from pipeLines

Adding information about retrieving host and thread into the MSOutputTemplate && Updating the documents in MongoDB repectively.

Generating Proper reports && Adding thred identificator to toplevel log messages.

Bug fix - messed up collection names && Exeption handling inside the pipeLines && DDM instance to __init__ && Few Important NOTES

Bug fix - missing quotes in paramenter list for docUploader

Update isRelval flag in the Producer pipeLine.

Allow 'None' values while constructing DDM requests && Fix an extra argument in ddm.makeRequest()

Switchin off setTaken flag in Consummer thread && Dump transfer request ouput for dryRun mode && Working on NonRelVals only.

Add info message in docDump function.

amaltaro commented:
@todor-ivanov please request for my review once you have fixed the problems reported by jenkins.

amaltaro commented:
PR adding pymongo dependency to our wmagent-dev spec/image:
https://github.com/cms-sw/cmsdist/pull/5874

todor-ivanov commented:
Test this please

amaltaro commented:
test this please

todor-ivanov committed:
pylint fixes

amaltaro commented:
While I review it, this must be solved:
```
Python3 compatibility checks: failed
    fails python3 compatibility test
```

amaltaro reviewed (changes requested):
Todor, I still haven't looked into a few blocks of code. In addition to the comments I made along the code, I have a few more questions/comments in general:
1) can you please remind me on the advantage of using bson.Timestamp instead an epoch timestamp?
2) what happens if MSOutput crashes or is restarted while it has documents "locked"?
3) Please remove `RequestStatus` from these documents. Having inconsistent information in different databases is only source of problems, inquiries and waste of time.
4) I'm not sure I read the code properly. It seems to fetch documents in bulk (according to the query string), right? When it updates or write documents into the database, is it writing documents one by one? Or we have bulk calls?

Thanks

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -151,20 +364,329 @@ def getRequestRecords(self, reqStatus):
+        of the document
         """
-        pass
+        return doc.clear()
```
No, but since at this stage (when I call this method) I am pretty sure there is nothing I need from these objects, so I'd rather clear them.

amaltaro replied:
My point is that cleaning them ourselves takes cpu cycles from our application. While leaving it up to gc, runs in the background. It's fine either way though.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -151,20 +364,329 @@ def getRequestRecords(self, reqStatus):
+        counter = 0
+        for _, request in requestRecords:
+            counter += 1
+            if 'SubRequestType' in request.keys() and 'RelVal' in request['SubRequestType']:
```
Good point.

todor-ivanov replied:
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -151,20 +364,329 @@ def getRequestRecords(self, reqStatus):
+        # Solution 3: To have 2 object buffers for the two diff types outside the function
+        try:
+            msOutDoc = MSOutputTemplate(doc)
+            doc.clear()
```
99% ....There is only one place I left a comment in the code about that. And i was hoping to hear you mentioning it and also your opinion: [1]
[1]
https://github.com/dmwm/WMCore/blob/4c1db77356de44bf987f7dd783d907a7f241d018/src/python/WMCore/MicroService/DataStructs/MSOutputTemplate.py#L156-L157

todor-ivanov replied:
Just to stay on the safe side I opted for the `deepcopy()`  here too. Which should resolve this comment too.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -151,20 +364,329 @@ def getRequestRecords(self, reqStatus):
+        """
+        A function used to update one or few particular fields in a document
+        :**kwargs: The keys/value pairs to be updated (will be tested against MSOutputTemplate)
+        """
```
Not a bad idea at all. I will add such a method to the class  now.

todor-ivanov replied:
I added the method to the class for future use, but for the moment I would like to keep the current way of updating the keys - First of all it is tested and I already accumulated quite a lot of new changes to be retested again. Second of all for me it is more readable if I can write:
```
docKeyUpdate(isTaken=False, 
                       isTakenby=None,
                       ...)
```

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -44,24 +59,70 @@ def __init__(self, msConfig, logger=None):
         self.msConfig.setdefault("rucioAccount", 'wma_test')
+        self.msConfig.setdefault("mongoDBUrl", 'mongodb://localhost')
+        self.msConfig.setdefault("mongoDBPort", 8230)
+        self.msConfig.setdefault("streamerBufferFile", "/tmp/msOutput/requestRecords")
```
Yes, it is for the moment how I feed the streamer. I would pretty much like to keep the parameter there though, but will put the default value to `None`. So that once the streamer fails to open the file it would immediately fall back to reading the requests from memory buffer (which have been passed to the constructor too [2], but for my tests was ignored because I provided the file artificially at the path in question).
[1]
https://github.com/dmwm/WMCore/blob/3a8ee05954b2c217f3d7080faf331d76ee82452e/src/python/WMCore/MicroService/Unified/MSOutputStreamer.py#L54-L62

[2]
https://github.com/dmwm/WMCore/blob/3a8ee05954b2c217f3d7080faf331d76ee82452e/src/python/WMCore/MicroService/Unified/MSOutput.py#L172

todor-ivanov replied:
Upon a second thought I can see your point. Having it configured here to point the proper place is good. But I am trying to avoid confusion until we fully develop this part. Because if the file we point to for some reason shows up into the system but is not the expected one it will substitute what ever is given in the memory buffer. I will keep it with `None` for the moment, but yes the correct path should be exactly:

`/data/srv/state/reqmgr2ms/ms-output/requestRecords` or `/data/srv/state/reqmgr2ms/ms-output/streamerBufferFile`

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -9,16 +9,45 @@
+from WMCore.MicroService.Unified.MSOutputStreamer import MSOutputStreamer
+
+
+class MSOutputException(Exception):
```
Done for the Parent Class. The other should inherit all the needed properties too.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -27,13 +56,21 @@ class MSOutput(MSCore):
     """
 
-    def __init__(self, msConfig, logger=None):
+    def __init__(self, msConfig, logger=None, mode=None):
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/DataStructs/DefaultStructs.py:
```diff
@@ -48,12 +48,22 @@
+                              total_num_requests=0,
+                              total_num_campaigns=0,
+                              num_datasets_subscribed=0,
+                              num_ddm_requests=0)
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/DataStructs/MSOutputTemplate.py:
```diff
@@ -0,0 +1,266 @@
+        docTemplate = [
+            ('_id', None, unicode),
+            ('RequestName', None, unicode),
+            ('RequestStatus', None, unicode),
```
Removed

todor-ivanov commented on src/python/WMCore/MicroService/DataStructs/MSOutputTemplate.py:
```diff
@@ -0,0 +1,266 @@
+            ('destination', None, list),
+            ('destinationOutputMap', None, dict),
+            ('campaignOutputMap', None, dict),
+            ('transferStatus', None, unicode),
```
Ok, this is something I am currently puzzled how to solve. And more precisely because of [1]. I am not feeling quite comfortable of casting things between `str` and `unicode` in the python code before checking them in the template.

[1]
https://api.mongodb.com/python/3.4.0/tutorial.html#a-note-on-unicode-strings

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSOutput.py:
```diff
@@ -151,20 +364,329 @@ def getRequestRecords(self, reqStatus):
+            # subscription have never been done for them
+            mQueryDict = {'isTaken': False, 'transferStatus': None}
+            try:
+                pipeLineName = msPipelineRelVal.name
```
Ok, Agreed. I am testing the result now.

todor-ivanov replied:
it is working now

todor-ivanov committed:
Second review iteration.

Second review second iteration.

Secon review Bugfix - Wrongly running twice the same pipiline

Secon review Bugfix - Bad exception class inheritance && Badly positioned thred identifier creation.

Second review third iteration - Add imports from future.

amaltaro commented on src/python/WMCore/MicroService/DataStructs/MSOutputTemplate.py:
```diff
@@ -0,0 +1,256 @@
+            ('lastUpdate', None, Timestamp),
+            ('isRelVal', None, bool),
+            ('isTaken', False, bool),
+            ('isTakenby', None, (str, unicode)),
```
Perhaps we should be consistent with these keys, so:
* "isTakenby" --> "isTakenBy"
* for the others, always start with a non-capital letter.

amaltaro replied:
then reflect the same changes in the docstring above

todor-ivanov replied:
> "isTakenby" --> "isTakenBy"
This I will fix. 

> for the others, always start with a non-capital letter.

It will take me out of consistency with the same keys coming from Reqmgr, because there they are all starting with Capital letters. And if I do so, I will need to implement additional logic to sort/search/map them while fetching them from Reqmgr Api and trying to modify them  to the current naming schema. While  right now it happens naturally because of this: [1]
I would rather say lets make them all starting with Capital letter here too.

[1]
https://github.com/dmwm/WMCore/blob/22297f258696f5d0704fc46517b4db2319d10e3f/src/python/WMCore/MicroService/DataStructs/MSOutputTemplate.py#L106-L108

todor-ivanov replied:
> "isTakenby" --> "isTakenBy"

Done

amaltaro replied:
>  I would rather say lets make them all starting with Capital letter here too.

I agree! Please start all of them with capital letters then.

todor-ivanov committed:
Use standard Unix epoch time (as int) instead of Bson Timestamp format && Avoid retrying workflows in the same polling cycle.

Typo

todor-ivanov committed:
Add Exception handling for the Producer thread.

todor-ivanov commented:
@amaltaro, Thank you for the review.  I think I addressed all the requested changes  to the level that is reasonable given how late is this PR. It has been tested in my VM and while working on the things you mentioned  I spotted few other errors which I also cleaned. Before we proceed further let me answer your last questions from the review:

> advantage of using bson.Timestamp instead an epoch timestamp

NONE :) It was  a mistake from my side to consider that since this is the internal format that Mongodb is using  we may benefit somehow from using it. It turned  out I cannot do simple arithmetics with it in the python code. So just went back to using Unix epoch time in seconds.

 > what happens if MSOutput crashes or is restarted while it has documents "locked"?

Documents stay locked. (of course it is in case we use the `isTaken` flag to lock them - which we agreed to switch off for the time being) 

> Please remove RequestStatus from these documents.

Removed

> It seems to fetch documents in bulk (according to the query string), right? When it updates or write documents into the database, is it writing documents one by one? Or we have bulk calls?

No, it is working in both directions a single document at a time. We can add calls to the database and fetch documents in bulk, but it will need some more changes in the code to make it run through the workflows in strides of some predefined size. We can do it, but it will require further development.

amaltaro commented:
@todor-ivanov changes look good in general, but I still see a few required changes here:
* first, on the mongo document key naming convention (once we start creating docs in mongodb, it's much harder to change anything...)
* you skipped my comments on the `MSOutputStreamer` module. Please review them and/or work on them in a new PR
* we definitely need unit tests for some of those modules (please create a new GH issue)
* please also create a GH issue to deal with those mongodb documents in bulk (that's an enhancement though)
